Content.Server/Ninja/Systems/NinjaSystem.cs

```diff
@@ -0,0 +1,572 @@
+    [Dependency] private readonly EmagSystem _emag = default!;
+    [Dependency] private readonly GameTicker _gameTicker = default!;
+    [Dependency] private readonly SharedHandsSystem _hands = default!;
+    [Dependency] private readonly SharedSubdermalImplantSystem _implants = default!;
+    [Dependency] private readonly IPrototypeManager _proto = default!;
+    [Dependency] private readonly PowerCellSystem _powerCell = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly TagSystem _tags = default!;
+    [Dependency] private readonly TraitorRuleSystem _traitorRule = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<ActivateInWorldEvent>(OnActivate);
+        SubscribeLocalEvent<NinjaDrainComponent, DoAfterEvent>(OnDrainDoAfter);
+        SubscribeLocalEvent<NinjaDownloadComponent, DoAfterEvent>(OnDownloadDoAfter);
+        SubscribeLocalEvent<NinjaTerrorComponent, DoAfterEvent>(OnTerrorDoAfter);
+
+        // TODO: maybe have suit activation stuff
+        SubscribeLocalEvent<SpaceNinjaSuitComponent, ContainerIsInsertingAttemptEvent>(OnSuitInsertAttempt);
+        SubscribeLocalEvent<SpaceNinjaSuitComponent, TogglePhaseCloakEvent>(OnTogglePhaseCloakAction);
+        SubscribeLocalEvent<SpaceNinjaSuitComponent, RecallKatanaEvent>(OnRecallKatanaAction);
+
+        SubscribeLocalEvent<SpaceNinjaComponent, ComponentStartup>(OnNinjaStartup);
+        SubscribeLocalEvent<SpaceNinjaComponent, MindAddedMessage>(OnNinjaMindAdded);
+        SubscribeLocalEvent<SpaceNinjaComponent, AttackedEvent>(OnNinjaAttacked);
+
+        SubscribeLocalEvent<DoorComponent, DoorEmaggedEvent>(OnDoorEmagged);
+    }
+
+    public override void Update(float frameTime)
+    {
+        foreach (var ninja in EntityQuery<SpaceNinjaComponent>())
+        {
+            var uid = ninja.Owner;
+            UpdateNinja(uid, ninja, frameTime);
+        }
+    }
+
+    /// <summary>
+    /// Turns the player into a space ninja
+    /// </summary>
+    public void MakeNinja(Mind.Mind mind)
+    {
+        if (mind.OwnedEntity == null)
+            return;
+
+        // prevent double ninja'ing
+        var user = mind.OwnedEntity.Value;
+        if (HasComp<SpaceNinjaComponent>(user))
+            return;
+
+        AddComp<SpaceNinjaComponent>(user);
+        SetOutfitCommand.SetOutfit(user, "SpaceNinjaGear", EntityManager);
+        GreetNinja(mind);
+    }
+
+    protected override void NinjaEquippedGloves(EntityUid uid, SpaceNinjaGlovesComponent comp, EntityUid user, SpaceNinjaComponent ninja)
+    {
+        base.NinjaEquippedGloves(uid, comp, user, ninja);
+
+        if (TryComp<ActionsComponent>(user, out var actions))
+            _actions.AddAction(user, comp.ToggleAction, uid, actions);
+    }
```

Comment:
You can just have this all on shared and use SharedActionSystem instead of just ActionsSystem. You can also replace the GotEquippedEvent with a GetItemActionsEvent since the latter handles items being picked up/equipped as well.

Then that way you can properly have a Toggle eventbus and use ToggleActionEvent to toggle and use a _actions.SetToggled(action, bool) to toggle it on and off properly.

Reply:
ok i am gonna do this it sounds smart

Reply:
i think i implemented this, but i dont use settoggled